src/VecSim/algorithms/brute_force/brute_force.h

```diff
@@ -83,3 +84,457 @@ class BruteForceIndex : public VecSimIndex {
+    void *timeoutCtx = queryParams ? queryParams->timeoutCtx : NULL;
+
+    this->last_mode = STANDARD_KNN;
+    DataType normalized_blob[this->dim]; // This will be use only if metric == VecSimMetric_Cosine.
+    if (this->metric == VecSimMetric_Cosine) {
+        // TODO: need more generic
+        memcpy(normalized_blob, queryBlob, this->dim * sizeof(DataType));
+        float_vector_normalize(normalized_blob, this->dim);
+        queryBlob = normalized_blob;
+    }
+
+    DistFuncType upperBound = std::numeric_limits<DistFuncType>::lowest();
+    vecsim_stl::max_priority_queue<pair<DistFuncType, labelType>> TopCandidates(this->allocator);
+    // For every block, compute its vectors scores and update the Top candidates max heap
+    idType curr_id = 0;
+    for (auto vectorBlock : this->vectorBlocks) {
+        auto scores = computeBlockScores(vectorBlock, queryBlob, timeoutCtx, &rl.code);
+        if (VecSim_OK != rl.code) {
+            return rl;
+        }
+        for (size_t i = 0; i < scores.size(); i++) {
+            // Always choose the current candidate if we have less than k.
+            if (TopCandidates.size() < k) {
+                TopCandidates.emplace(scores[i], getVectorLabel(curr_id));
+                upperBound = TopCandidates.top().first;
+            } else if (scores[i] < upperBound) {
+                // Otherwise, try greedily to improve the top candidates with a vector that
+                // has a better score than the one that has the worst score until now.
+                TopCandidates.emplace(scores[i], getVectorLabel(curr_id));
+                TopCandidates.pop();
+                upperBound = TopCandidates.top().first;
+            }
+            ++curr_id;
+        }
+    }
+    assert(curr_id == count);
+
+    rl.results = array_new_len<VecSimQueryResult>(TopCandidates.size(), TopCandidates.size());
+    for (int i = (int)TopCandidates.size() - 1; i >= 0; --i) {
+        VecSimQueryResult_SetId(rl.results[i], TopCandidates.top().second);
+        VecSimQueryResult_SetScore(rl.results[i], TopCandidates.top().first);
+        TopCandidates.pop();
+    }
+    rl.code = VecSim_QueryResult_OK;
+    return rl;
+}
+
+template <typename DataType, typename DistFuncType>
+VecSimQueryResult_List
+BruteForceIndex<DataType, DistFuncType>::rangeQuery(const void *queryBlob, DistFuncType radius,
+                                                    VecSimQueryParams *queryParams) {
+    auto rl = (VecSimQueryResult_List){0};
+    void *timeoutCtx = queryParams ? queryParams->timeoutCtx : nullptr;
+    this->last_mode = RANGE_QUERY;
+
+    DataType normalized_blob[this->dim]; // This will be use only if metric == VecSimMetric_Cosine.
+    if (this->metric == VecSimMetric_Cosine) {
+        // TODO: need more generic when other types will be supported.
+        memcpy(normalized_blob, queryBlob, this->dim * sizeof(DataType));
+        float_vector_normalize(normalized_blob, this->dim);
```

Comment:
Same here (is making it generic should be part of this PR? + avoid allocating on stack if not have to)